My Site Dashboard: Phase 2: Metrics - Track Cards Shown

zwarm opened the pull request:
Fixes #15732 

This PR implements shown tracking as follows:
- **MY_SITE_DASHBOARD_CARD_SHOWN** event for dashboard "post" cards with the following properties
     - type = “post”
     - sub_type = “create_first”, “create_next”, “drafts”, “scheduled”

- Tracking for Dashboard ErrorCard
**MY_SITE_DASHBOARD_CARD_SHOWN** event with the following properties
     - type = “error”

- Moved the **DOMAIN_CREDIT_PROMPT_SHOWN** event from `CardsBuilder` to `CardsShownTracker`.
- Added helper class `CardsShownTracker`  
- Tracking whether or not a shown card event was tracked is encapsulated in `CardsTracker` (for dashboard cards) and in `CardsShownTracker` (for non-dashboard cards).

**MERGE INSTRUCTIONS**
I put the **Not Ready For Merge** label on this to leave room for any possible direction changes. If there are no direction changes, then please remove the label and treat it as a normal PR. Thanks. 🙇 


**Notes**:
- Cards shown are tracked 1x for each site
- Cards shown events are not tracked onResume or return from background (unless the OS kills the process)
- Cards shown events are cleared when the selected site changes
- If a user is shown the create first post card, then creates a post, we will track create next after it is shown. Same goes for drafts or scheduled. Each card can trigger a shown 1x per site change. 
- Tapping the site selector and not changing the site, will not trigger a new tracking.
- The domain registration card is shown tracking event will follow the same pattern as the dashboard cards. Previously this event was tracked during onResume. We should note this in the tracks posts. 

**Other implementation Notes**
- I was going back-n-forth on naming the `CardsShownTracker`. I was going to name it `CardsTracker` and move other card tracking events to it, but there wasn't anything to add Also I didn't like the name conflict with `CardsTracker` under the dashboard package. As always, I am open to changing names.
- I thought about changing the way "domain" shown is tracked; using a **MY_SITE_CARD_SHOWN**  pattern that mimics the **MY_SITE_DASHBOARD_CARD_SHOWN**; however I didn't want to throw off any funnels that depend on it. We could "double" track that with the old event name + implement a new way. Would love thoughts on that?
- I went with this implementation because we are always recreating the UiModel/State fresh (no 'copy") . I couldn't use partial state, because that occurs before the card is built. If you can think of another implementation, I am more than open to switching. Tried to keep it simple and to the point.
- I thought about using **MY_SITE_DASHBOARD_ERROR_CARD_SHOWN**, but went with the type property instead. In the future, when we are tracking individual "card" errors, we can easily introduce a subtype. 
- I did not track Posts as `PostCardWithItems` or `PostCardWithoutItems`, I felt that "post" along with subtype was the better option (it's self explanatory)
 
I know this was a lot of chatter, but thanks for making it to the bottom. This 🤗  is for you!!

**To test:**
Setup
- Launch the app
- Navigate to Me -> App Settings -> Debug Settings
- Ensure that the `MySiteDashboardPhase2FeatureConfig` is to set to on
- Restart the app if needed
- Navigate to Me -> App Settings -> Privacy Settings 
- Toggle the "collect information" switch to on
------------------------
- Navigate to the My Site tab
- Use the Site selector to chose a site that has no posts
- Navigate to Me -> Help and Support -> Application log
- Note that the following is shown in the log
   I: 🔵 Tracked: my_site_dashboard_card_shown, Properties: {"type":"post","subtype":"create_first"}
-----------------------
- Navigate to the My Site tab
- Use the Site selector to chose a site has posts, but no draft or scheduled posts
- Navigate to Me -> Help and Support -> Application log
- Note that the following is shown in the log
I: 🔵 Tracked: my_site_dashboard_card_shown, Properties: {"type":"post","subtype":"create_next"}
-----------------------
- Navigate to the My Site tab
- Use the Site selector to chose a site that has draft posts
- Navigate to Me -> Help and Support -> Application log
- Note that the following is shown in the log
I: 🔵 Tracked: my_site_dashboard_card_shown, Properties: {"type":"post","subtype":"draft"}
-----------------------
- Navigate to the My Site tab
- Use the Site selector to chose a site that has scheduled posts
- Navigate to Me -> Help and Support -> Application log
- Note that the following is shown in the log
I: 🔵 Tracked: my_site_dashboard_card_shown, Properties: {"type":"post","subtype":"scheduled"}
---------------------

I couldn't trigger the error card in the app, so if to test you need to pull down this PR and adjust CardsBuilder.build to only create the `ErrorCard`. Run the app, navigate to My Site and then check the application log for :
I: 🔵 Tracked: my_site_dashboard_card_shown, Properties: {"type":"error"}

------------------
- Using any of the testing scenarios above (excluding error, unless you feel like setting up the logic)
- Note the application log has tracked card shown
- Turn your device to trigger a configuration change
- Navigate to application log
- Note the event didn't trigger again 

## Regression Notes
1. Potential unintended areas of impact
Cards shown are not properly tracked

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual and automated tests

3. What automated tests I added (or what prevented me from doing so)
Add tests to MySiteViewModel, CardsTracker, CardsShownTracker.

PR submission checklist:

- [X] I have completed the Regression Notes.
- [X] I have considered adding accessibility improvements for my changes.
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

zwarm committed:
Remove tracking for domain registration card in the builder, as with the source logic, this is tracking the card multiple times for a single load. Tracking will be invoked in MySiteViewModel in a later commit.

zwarm committed:
Add support for tracking MY_SITE_DASHBOARD_CARD_SHOWN

zwarm committed:
Refactor: rename Subtype to PostSubtype

zwarm committed:
Refactor: change parameter types so the function can be used to track any card type/subtype

zwarm committed:
Add helper class for managing card shown tracking events

zwarm committed:
Implement cardsShownTracker for handling tracking shown card events for like DomainRegistration. Excludes dashboardCards and siteItems.

zwarm committed:
Add cardsShownTracker mock so that test don't fail. Tests for the tracker will be added in a later commit.

zwarm committed:
Add support for clearing the cardsShownTracker when the selected site has changed.

zwarm committed:
Add support for tracking post cards shown. The error card will be implemented in the next PR

zwarm committed:
Add support for tracking error cards shown.

zwarm committed:
Reset error card shown when requested.

zwarm committed:
Refactor: Replaced domain registration tracking test with tests for domain registration card. The tracking test has been moved to MySiteViewModelTest (future commit).

zwarm committed:
Add tests for domain registration and dashboard card shown track event verification

zwarm committed:
Add tests for verifying shown trackers are reset when selected site changes.

zwarm committed:
Add tests for CardsShownTracker

zwarm committed:
Add tests for cards shown

zwarm committed:
Address ktlint issues

zwarm committed:
Refactor: streamline the buildDashboardCards fun

ParaskP7 commented:
👋 @zwarm !

> Moved the DOMAIN_CREDIT_PROMPT_SHOWN event from CardsBuilder to CardsShownTracker.
> 
> I thought about changing the way "domain" shown is tracked; using a MY_SITE_CARD_SHOWN pattern that mimics the MY_SITE_DASHBOARD_CARD_SHOWN; however I didn't want to throw off any funnels that depend on it. We could "double" track that with the old event name + implement a new way. Would love thoughts on that?

I am kind of confused about these two sentences and what I am seeing being implemented, let me explain:
- By moving the `DOMAIN_CREDIT_PROMPT_SHOWN` from `CardsBuilder` to `CardsShownTracker` we effectively changing the tracking behavior. This tracking event will no longer get tracked during on resume, or as it was tracked before, right?
- I now actually agree with your thinking of not changing the way "domain" shown is tracked so that to not affect any funnels, thus if the above is true, maybe it is better if we revert it? I am also seeing the change within the commits, so I guess the above is true. PS: Apologies for not thinking that before, during our talk yesterday and giving the 👍 to proceed with this change.
- At this point, and so that we keep it simple, I wouldn't "double" track the old event name + implement a new way. Actually that is a great idea, but if we go this path, I would like us to implement a shown tracking event for all cards, thus I would leave all this for now and get back to it later on when we get enough data on how the dashboard card shown tracking goes. Maybe it will end up not helping us. Maybe the current implementation will provide false positives or negatives. Let's first wait for a couple of releases to gather some data and then plan for any enhancements. Wdyt?

My 2' cents... 😅

zwarm commented:
> * By moving the `DOMAIN_CREDIT_PROMPT_SHOWN` from `CardsBuilder` to `CardsShownTracker` we effectively changing the tracking behavior. This tracking event will no longer get tracked during on resume, or as it was tracked before, right?

Just to make this clear in the ticket,  the `DOMAIN_CREDIT_PROMPT_SHOWN` event is being tracked multiple times, including return from background. There is no explicit onResume() calls to track. However with the architecture changes introduced for phase 2 of msd, leaving the tracking in CardsBuilder results in the event getting tracked 5-6x more than it's being tracked now. To address this inconsistency, we are adjusting the shown track event to align with what we are doing for the dashboard cards, as it will more accurately reflect the truth.

Hope this helps. :)

zwarm committed:
Refactor: rename CardsShowTracker to DomainRegistrationCardShownTracker.

zwarm committed:
Extract card shown logic out of CardsTracker in to a separate class to make it easier to replace if needed.

zwarm committed:
Refactor: Extracted card shown logic into CardsShownTracker

zwarm committed:
Refactor: rename trackCardShown to trackShown to stay consistent with CardsTracker.

zwarm committed:
Updated card shown tests

zwarm committed:
Refactor: Adjust where trackCardsAndItemsShownIfNeeded is called in the state property, so we only call it on SiteSelected state.

zwarm committed:
Refactor: Removed cards shown tests to new class (next commit)

zwarm committed:
Moved cards shown test from CardsTracker and adjusted where needed.

zwarm committed:
Address ktlint issues

zwarm committed:
Merge with develop and resolve conflicts

zwarm commented:
@ParaskP7 - I've updated the PR to include
- Renamed the original `CardsShownTracker` to `DomainRegistrationCardsShownTracker`. The logic remains the same, yes, even though we are just passing DomainRegistration cards. This way it will be easier to rename/reuse in the future.
- Extracted card shown logic out of `CardsTracker` and added to new `CardsShownTracker`
- Extracted card shown logic out of `CardsTrackerTest` and added to new `CardsShownTrackerTest`
- MySiteViewModel - I used some of the patch (the state part), but left the `trackCardsAndItemsShownIfNeeded` and `resetShownTrackers` functions because they need to invoke both trackers (domain & dashboard cards).
- Updated MySiteViewModelTest
- Merged in develop to resolve conflicts due to the addition of onRetryClick of the ErrorCard

I think that should take care of all we talked about this morning. I have removed the not ready for merge label. Please take another look when you have the chance. 🙏  🙇

ParaskP7 reviewed (approved):
👋 @zwarm !

I have reviewed and tested this PR as per the instructions, everything works as expected, excellent work! 🌟 🌟 🌟 🎄
